File: src/main/java/es/udc/fi/dc/fd/model/services/RoutineService.java
package es.udc.fi.dc.fd.model.services;

import java.util.List;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import es.udc.fi.dc.fd.model.common.exceptions.DuplicateInstanceException;
import es.udc.fi.dc.fd.model.common.exceptions.InstanceNotFoundException;
import es.udc.fi.dc.fd.model.entities.Exercise;
import es.udc.fi.dc.fd.model.entities.Routine;
import es.udc.fi.dc.fd.model.entities.Serie;
import es.udc.fi.dc.fd.model.entities.Training;
import es.udc.fi.dc.fd.model.entities.Users;
import es.udc.fi.dc.fd.model.services.exceptions.InvalidRoutineDurationException;
import es.udc.fi.dc.fd.model.services.exceptions.InvalidRoutineNameException;
import es.udc.fi.dc.fd.model.services.exceptions.PermissionException;
import es.udc.fi.dc.fd.model.services.exceptions.RoutineExerciseLimitReachedException;
import es.udc.fi.dc.fd.model.services.exceptions.RoutineLimitReachedException;

public interface RoutineService {
    Routine createRoutine(Long creatorId, String name, List<Long> exercises, Long duration, Boolean isPublic) throws DuplicateInstanceException, InstanceNotFoundException, 
    InvalidRoutineNameException, InvalidRoutineDurationException, RoutineLimitReachedException, RoutineExerciseLimitReachedException;
    
    Page<Routine> viewAllRoutines(Long userId, Pageable pageable) throws InstanceNotFoundException;

    Routine getRoutineById(Long routineId, Long userId) throws InstanceNotFoundException, PermissionException;
    
    Routine modifyRoutine(Long routineId, Long creatorId, String name, List<Long> exercises, Long duration, Boolean isPublic) throws InstanceNotFoundException, PermissionException, RoutineExerciseLimitReachedException;

    boolean deleteSeriesByRoutine(Long routineId) throws InstanceNotFoundException;

    void deleteRoutine(Long creatorId, Long routineId) throws InstanceNotFoundException, PermissionException;

    Page<Routine> findByFilters(Long userId, Long creatorId, String name, Pageable pageable) throws InstanceNotFoundException;

    void createTraining(Long userId, String trainingName, String trainingDescription, Boolean isPublic) throws InstanceNotFoundException;

    List<Serie> getDefaultRoutineSeries(Long routineId, Long exerciseId) throws InstanceNotFoundException;

    Training createTrainingFromRoutine(Long userId, String trainingName, String trainingDescription, Long duration, Boolean isPublic, List<Serie> series, Long routineId) throws InstanceNotFoundException;

    boolean followRoutine(Long userId, Long routineId) throws InstanceNotFoundException, PermissionException;

    boolean unfollowRoutine(Long userId, Long routineId) throws InstanceNotFoundException;

    Block<Users> getFollowersByRoutine(Long routineId, Long trainerId, Pageable pageable) throws InstanceNotFoundException, PermissionException;

    Page<Training> findTrainings(Long userId, Long id, Pageable pageable) throws InstanceNotFoundException, PermissionException;

    Page<Training> findTrainingsByDay(Long userId, int day, int month, int year, Pageable pageable) throws InstanceNotFoundException, PermissionException;

    List<Exercise> findTrainingExercises(Long trainingId) throws InstanceNotFoundException;

    List<Training> findTrainingsByYear(Long userId, int year) throws InstanceNotFoundException, PermissionException;

    Routine getRoutineByTraining(Long trainingId) throws InstanceNotFoundException;

    Training findTrainingById(Long trainingId) throws InstanceNotFoundException;
}

File: frontend/src/backend/routineService.js
import { fetchConfig, appFetch } from "./appFetch";

export const createRoutine = (name, exercises, duration, isPublic, onSuccess, onErrors) => {
  appFetch(
    "/routines/createRoutine",
    fetchConfig("POST", { name, exercises, duration, isPublic }),
    onSuccess,
    onErrors
  );
};

export const findRoutineById = (routineId, onSuccess, onErrors) =>
  appFetch(
    `/routines/getRoutineById/${routineId}`,
    fetchConfig("GET"),
    onSuccess,
    onErrors
  );

export const findRoutineDetails = (routineId, onSuccess, onErrors) =>
  appFetch(
    `/routines/getRoutineDetailsById/${routineId}`,
    fetchConfig("GET"),
    onSuccess,
    onErrors
  );

export const viewAllRoutines = ({ page, size }, onSuccess, onErrors) =>
  appFetch(
    `/routines/viewAllRoutines?page=${page}&size=${size}`,
    fetchConfig("GET"),
    onSuccess,
    onErrors
  );

export const deleteRoutine = (routineId, onSuccess, onErrors) =>
  appFetch(
    `/routines/deleteRoutine/${routineId}`,
    fetchConfig("DELETE"),
    onSuccess,
    onErrors
  );

export const modifyRoutine = (routineId, name, exercises, duration, isPublic, onSuccess, onErrors) =>
  appFetch(
    `/routines/modifyRoutine/${routineId}`,
    fetchConfig("PUT", { name, exercises, duration, isPublic }),
    onSuccess,
    onErrors
  );

export const searchRoutines = (creatorId, name, { page, size }, onSuccess, onErrors) =>
  appFetch(
    `/routines/search?creatorId=${creatorId || ""}&name=${name || ""}&page=${page}&size=${size}`,
    fetchConfig("GET"),
    onSuccess,
    onErrors
  );

export const createTraining = (
  { routineId, name, description, duration, visibility, exercises },
  { onSuccess, onErrors } = {}
) => {
  appFetch(
    "/routines/createTraining",
    fetchConfig("POST", { routineId, name, description, duration, visibility, exercises }),
    onSuccess,
    onErrors
  );
};

export const followRoutine = (routineId, onSuccess, onErrors) => {
  appFetch(
    `/routines/${routineId}/follow`,
    fetchConfig("POST"),
    onSuccess,
    onErrors
  );
};

export const unfollowRoutine = (routineId, onSuccess, onErrors) => {
  appFetch(
    `/routines/${routineId}/unfollow`,
    fetchConfig("DELETE"),
    onSuccess,
    onErrors
  );
};

export const getFollowersByRoutine = (routineId, { page, size }, onSuccess, onErrors) => {
  appFetch(
    `/routines/${routineId}/followers?page=${page}&size=${size}`,
    fetchConfig("GET"),
    onSuccess,
    onErrors
  );
};

export const isFollowingRoutine = (routineId, onSuccess, onErrors) => {
  appFetch(
    `/routines/${routineId}/is-following`,
  );
};

export const getTrainingDetails = (trainingId, onSuccess, onErrors) => {
  appFetch(
    `/routines/trainings/${trainingId}/details`,
    fetchConfig("GET"),
    onSuccess,
    onErrors
  );
}

export const viewUserTrainings = (page, size, onSuccess, onErrors) => {
  appFetch(
    `/routines/findTrainings?page=${page}&size=${size}`,
    fetchConfig("GET"),
    onSuccess,
    onErrors
  );
};

export const viewDayTrainings = (page, size, day, month, year, onSuccess, onErrors) => {
  appFetch(
    `/routines/findDayTrainings?day=${day}&month=${month}&year=${year}&page=${page}&size=${size}`,
    fetchConfig("GET"),
    onSuccess,
    onErrors
  );
};


export const getTrainingCalendarStats = (year, onSuccess, onErrors) => {
  appFetch(
    `/routines/getTrainingCalendarStats?year=${year}`,
    fetchConfig("GET"),
    onSuccess,
    onErrors
  );
};


